titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java

```diff
@@ -0,0 +1,151 @@
+                  Function<E, C> converter,
+                  Supplier<Predicate<E>> expiredPredicateSupplier,
+                  GrpcObjectsCacheConfiguration configuration,
+                  TitusRuntime titusRuntime) {
+        this.converter = converter;
+        this.expiredPredicateSupplier = expiredPredicateSupplier;
+        this.configuration = configuration;
+        this.clock = titusRuntime.getClock();
+
+        Registry registry = titusRuntime.getRegistry();
+        PolledMeter.using(registry).withId(registry.createId(METRICS_ROOT + "cacheSize", "cache", name)).monitorSize(cache);
+        PolledMeter.using(registry).withId(registry.createId(METRICS_ROOT + "toBeRemoved", "cache", name)).monitorSize(toBeRemoved);
+        this.cacheHit = registry.counter(METRICS_ROOT + "cacheAccess", "cache", name, "cacheHit", "true");
+        this.cacheMiss = registry.counter(METRICS_ROOT + "cacheAccess", "cache", name, "cacheHit", "false");
+
+        ScheduleDescriptor gcScheduleDescriptor = ScheduleDescriptor.newBuilder()
+                .withName(ProtobufCache.class.getSimpleName() + "#" + name)
+                .withDescription("Protobuf entity cache")
+                .withInitialDelay(Duration.ofMillis(configuration.getCacheRefreshInitialDelayMs()))
+                .withInterval(Duration.ofMillis(configuration.getCacheRefreshIntervalMs()))
+                .withTimeout(Duration.ofMillis(configuration.getCacheRefreshTimeoutMs()))
+                .build();
+        this.scheduleRef = titusRuntime.getLocalScheduler().schedule(gcScheduleDescriptor, context -> removeExpiredEntries(), true);
+    }
+
+    void shutdown() {
+        scheduleRef.cancel();
+    }
+
+    C get(String id, E entity) {
+        CacheEntry<E, C> entry = cache.get(id);
+        if (entry != null) {
+            if (entry.getCoreEntity() == entity) {
+                cacheHit.increment();
+                return entry.getProtobufEntity();
+            }
+        }
+        CacheEntry<E, C> newEntry = new CacheEntry<>(id, entity, converter.apply(entity));
+        cache.put(id, newEntry);
+        cacheMiss.increment();
+        return newEntry.getProtobufEntity();
+    }
+
+    private void removeExpiredEntries() {
+        long now = clock.wallTime();
+
+        // Do cleanup
+        Iterator<Map.Entry<String, Long>> it = toBeRemoved.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<String, Long> entry = it.next();
+            if (entry.getValue() < now) {
+                it.remove();
+                cache.remove(entry.getKey());
+            }
+        }
+
+        // Find expired entries
+        Predicate<E> expiredPredicate = expiredPredicateSupplier.get();
+        Set<String> candidates = CollectionsExt.copyAndRemove(cache.keySet(), toBeRemoved.keySet());
+        candidates.forEach(id -> {
+            CacheEntry<E, C> entry = cache.get(id);
+            if (expiredPredicate.test(entry.getCoreEntity())) {
```

Comment:
there's quite a bit of re-work here, since the predicate for each entity will recompute the `all[Job|Task]Ids` `HashMap` for entry. It's probably worth changing the API here to compute `items that should be present` only once to be tested against all entries in the cache.

Alternatively, change to capacity based eviction (instead of periodic time-based cleanup) as I mentioned on a previous comment.

Reply:
We want to minimize calls to JobService API which are relatively expensive. So before checking the expiry of cached entries, we fetch all of them with a single call. Next, we use the cached value (in the clojure), to check every item (this is fast).

Reply:
Yeah, I get the intent, let me rephrase for clarity. The `expiredPredicate` used here (reproduced below) is evaluated for each candidate (all cache entries), and each time it is evaluated, it recomputes the `known[Job|Task]Ids` Set:

```diff
+                () -> {
+                    List<com.netflix.titus.api.jobmanager.model.job.Job> allJobs = jobOperations.getJobs();
+                    Set<String> knownJobIds = new HashSet<>();
+                    allJobs.forEach(job -> knownJobIds.add(job.getId()));
+                    return job -> !knownJobIds.contains(job.getId());
+                },
+
```

I know `jobOperations.getJobs()` is all in-memory and is a lightweight operation, and building a `HashSet` out of it is not expensive. It's probably not a bottleneck, but removing the O(N^2) here and computing `known[Job|Tasks]Ids` set only once every eviction cycle would be a simple enough optimization that is probably worth it.

Reply:
My bad, I didn't realize we were holding a `Supplier<Predicate>` here, and computing the whole `Predicate` only once before checking each item. Nevermind my comments, LGTM 👍